File: 0070_climb_stairs/climb_stairs.go
package main

/**
 * Time: O(n)
 * Space: O(n)
 */
func climbStairs(n int) int {
	if n == 1 {
		return 1
	}
	dp := make([]int, n)
	dp[0], dp[1] = 1, 2
	for i := 2; i < n; i++ {
		dp[i] = dp[i-2] + dp[i-1]
	}
	return dp[n-1]
}

func main() {

}

File: 0070_climb_stairs/climb_stairs.py
def climb_stairs(n: int) -> int:
    """
    Time: O(n)
    Space: O(n)
    """
    memo = {}
    return helper(n, 0, memo)


def helper(n: int, curr: int, memo) -> int:
    if curr > n:
        return 0
    if curr == n:
        return 1
    if curr in memo:
        return memo[curr]
    memo[curr] = helper(n, curr + 1, memo) + helper(n, curr + 2, memo)
    return memo[curr]
